src/flash.rs

```diff
@@ -0,0 +1,165 @@
+use crate::signature::FlashSize;
+use crate::stm32::FLASH;
+use core::{ptr, slice};
+
+/// Flash erase/program error
+#[derive(Debug, Clone, Copy)]
+pub enum Error {
+    ProgrammingSequence,
+    ProgrammingParallelism,
+    ProgrammingAlignment,
+    WriteProtection,
+    Operation,
+}
+
+impl Error {
+    fn read(flash: &FLASH) -> Option<Self> {
+        let sr = flash.sr.read();
+        if sr.pgserr().bit() {
+            Some(Error::ProgrammingSequence)
+        } else if sr.pgperr().bit() {
+            Some(Error::ProgrammingParallelism)
+        } else if sr.pgaerr().bit() {
+            Some(Error::ProgrammingAlignment)
+        } else if sr.wrperr().bit() {
+            Some(Error::WriteProtection)
+        } else if sr.operr().bit() {
+            Some(Error::Operation)
+        } else {
+            None
+        }
+    }
+}
+
+/// Flash methods implemented for `stm32::FLASH`
+pub trait FlashExt {
+    /// Memory-mapped address
+    fn address(&self) -> usize;
+    /// Size in bytes
+    fn len(&self) -> usize;
+    /// Returns a read-only view of flash memory
+    fn read(&self) -> &[u8] {
+        let ptr = self.address() as *const _;
+        unsafe { slice::from_raw_parts(ptr, self.len()) }
+    }
+    /// Unlock flash for erasing/programming until this method's
+    /// result is dropped
+    fn unlocked(&mut self) -> UnlockedFlash;
+}
+
+impl FlashExt for FLASH {
+    fn address(&self) -> usize {
+        0x0800_0000
```

Comment:
Could this be a static fn instead? Same question applies to `len()`.

Reply:
`read` needs to borrow `self`, otherwise the user would be able to have a `&` and `&mut` to the same memory.

Reply:
Sorry, not `&mut`, but they would be able to mutate the memory while it has a live immutable reference to it.

Reply:
> read needs to borrow self, otherwise the user would be able to have a & and &mut to the same memory.
> Sorry, not `&mut`, but they would be able to mutate the memory while it has a live immutable reference to it.

I don't think you're carrying around a reference to memory, though; only the results of a read (i.e. copied bytes). I believe I see what you mean if you think of the entirety of Flash as an object to which you have a slice into, but I don't think very much is gained from treating it that way. I'm really curious to hear what you/others think about it.  I do see how this could become a problem in environments where two threads could simultaneously access flash, though.

Reply:
> I see what you mean if you think of the entirety of Flash as an object to which you have a slice into, but I don't think very much is gained from treating it that way.

`read` is doing exactly that, it gives the user a `&[u8]` pointing directly to flash memory. Without the borrow, the user would be able to create an `UnlockedFlash` and modify the flash contents, and then use the same `&[u8]` after, which is UB in rust, since you can't mutable alias with references.

Reply:
`read` doesn't copy. But even if it did (which isn't really desirable), without borrowing `Self` you couldn't prevent the user from using `read` on main and `program` in an interrupt, which is a data race.

Reply:
You are right -- I was thinking of my own implementation for `read()`, not the one included here :man_facepalming: .  Apologies.